vm/src/builtins/type.rs

```diff
@@ -1053,7 +1050,12 @@ impl Constructor for PyType {
         let heaptype_member_count = heaptype_slots.as_ref().map(|x| x.len()).unwrap_or(0);
         let member_count: usize = base_member_count + heaptype_member_count;
 
-        let flags = PyTypeFlags::heap_type_flags() | PyTypeFlags::HAS_DICT;
+        let mut flags = PyTypeFlags::heap_type_flags();
+        // Only add HAS_DICT and MANAGED_DICT if __slots__ is not defined.
+        if heaptype_slots.is_none() {
+            flags |= PyTypeFlags::HAS_DICT | PyTypeFlags::MANAGED_DICT;
+        }
+
```

Comment:
How about splitting the patch one to contains this part and the other one contains the test_class part? I'd like to merge the test_class part as the same form of CPython patch, but this one doesn't need to be.

Reply:
I'll create a new PR for the implementation part.